src/Compiler/Postprocessor.php

```diff
@@ -30,26 +37,61 @@ final class Postprocessor
 
     private CompilerInterface $compiler;
 
+    private ScopeFactory $scopeFactory;
+
+    private NodeScopeResolver $nodeScopeResolver;
+
     public function __construct(
         Parser $parser,
         NodeVisitorStorage $nodeVisitorStorage,
         Standard $printerStandard,
-        CompilerInterface $compiler
+        CompilerInterface $compiler,
+        ScopeFactory $scopeFactory,
+        NodeScopeResolver $nodeScopeResolver
     ) {
         $this->parser = $parser;
         $this->nodeVisitorStorage = $nodeVisitorStorage;
         $this->printerStandard = $printerStandard;
         $this->compiler = $compiler;
+        $this->scopeFactory = $scopeFactory;
+        $this->nodeScopeResolver = $nodeScopeResolver;
     }
 
-    public function postProcess(string $phpContent, Template $template): string
+    public function postProcess(string $phpContent, Template $template, string $compileFilePath): string
     {
         $phpStmts = $this->findNodes($phpContent);
         foreach ($this->nodeVisitorStorage->getNodeVisitors() as $nodeVisitors) {
             $phpStmts = $this->processNodeVisitors($phpStmts, $nodeVisitors, $template);
         }
 
-        return $this->printerStandard->prettyPrintFile($phpStmts);
+        $phpContent = $this->printerStandard->prettyPrintFile($phpStmts);
+        file_put_contents($compileFilePath, $phpContent);
+        $realPath = realpath($compileFilePath) ?: '';
+        if ($realPath === '') {
+            return '';
+        }
+        require($compileFilePath); // load type definitions from compiled template
+
+        $scope = $this->scopeFactory->create(ScopeContext::create($realPath));
+
+        $phpStmts = $this->findNodes($phpContent);
+
+        $nodeScopeResolver = clone $this->nodeScopeResolver;
+        $nodeScopeResolver->processNodes($phpStmts, $scope, function (Node $node, Scope $scope) {
+            if ($node instanceof Expr && !$node instanceof EncapsedStringPart) {
+                $node->setAttribute(ExprTypeNodeVisitorInterface::TYPE_ATTRIBUTE, $scope->getType($node));
+            }
+        });
+
+        foreach ($this->nodeVisitorStorage->getNodeVisitors(true) as $nodeVisitors) {
```

Comment:
so first all non-scope aware visitors are called and then al scope aware? Is there reason for that?

Reply:
Yes, first we run all visitors which don’t need scope. But main reason is how phpstan scope works. We have to store content to file, load that file and also require it because phpstan want to know all symbols - classes. So we can require this file only once because multiple require on same file will throw error. That's also why we can’t create it as service and run it “on demand”. This is what we have to document - scope aware node visitors can’t add new classes because it will not work then.